### DISCUSSION TITLE
Notes on upgrading to Ubuntu 24.04 LTS (on WSL)

### DISCUSSION THREAD

**Discussion by BJReplay** [jasonacox/Powerwall-Dashboard#510] [General]
Hi All

I was a little bored the other day (OK, I had stuff to do, and was procrastinating), so I thought I would bite the bullet and update a bunch of machines from Ubuntu LTS 22.04 to Ubuntu LTS 24.04.

What could go wrong?

As it turns out, quite a bit, but not that much that I was off the air for too long.

Thanks to .bash_history, here are the notes of what I had to do to get Powerwall Dashboard up and running on Ubuntu LTS 24.04 running on WSL (without Docker Desktop) - instructions for that - which I had to refer to - are at https://github.com/jasonacox/Powerwall-Dashboard/blob/main/WINDOWS.md

The key problems I ran into were:
- Python virtual environments are strictly enforced, so all the tools like Tesla History Tool (https://github.com/jasonacox/Powerwall-Dashboard/blob/main/tools/tesla-history/README.md), the mysql connector calculation update script (https://github.com/jasonacox/Powerwall-Dashboard/tree/main/tools/mysql#readme) needed to be re-set-up, including
  - installing python virtual environment support
  - creating a python virtual environment
  - downloading prerequisites
- for some reason, the docker install didn't upgrade, so I had to re-install docker (but once re-installed, everything re-started)

Total downtime for the collector was about 15 minutes that the history tool backfilled for me.

The upgrade process is pretty well covered elsewhere, but in summary, the steps that I did (modified, because I'm running WSL, so where you see reboot, substitute that with `wsl --terminate Ubuntu` followed by `wsl --shutdown` to shut down the WSL subsystem, instead of `sudo reboot`) are pretty straight forward:
- Update everything first.
- Clean up redundant packages
- Reboot
- Do the upgrade
- Update everything again.
- Clean up redundant packages
- Set up Python Virtual Environment
- Install Python Prerequisites for the tools you need (you may need more - error messages will point you in the right direction)
- May need to re-set-up docker (I only needed to do this on WSL)
- May need to re-add custom repos (e.g. Tailscale)
- Backfill gaps using Telsa History tool

Detailed Steps:

1. Update `sudo apt update && sudo apt dist-upgrade -y`
2. Remove redundant packages `sudo apt autoremove`
3. Reboot `sudo reboot`
4. Confirm that upgrades are set to LTS `cat /etc/update-manager/release-upgrades` and look for `prompt=lts`
5. Do the upgrade `sudo do-release-upgrade -d`
6. Update packages again post upgrade `sudo apt update && sudo apt dist-upgrade -y`
7. Remove redundant packages `sudo apt autoremove`
8. Reboot `sudo reboot`
9. Create Python virtual environment directory (I created mine under my Powerwall-Dashboard directory, but can be anywhere): `cd Powerwall-Dashboard && mkdir venv`
10. Install Python Virtual Environment support `sudo apt install python3.12-venv`
11. Create Python virtual environment `python3 -m venv ~/Powerwall-Dashboard/venv`
12. Install necessary Python components and libraries (first two for Tesla History, remainder for MySQL Support):
```
~/Powerwall-Dashboard/venv/bin/pip install teslapy
~/Powerwall-Dashboard/venv/bin/pip install python-dateutil
~/Powerwall-Dashboard/venv/bin/pip install influxdb
~/Powerwall-Dashboard/venv/bin/pip install sqlalchemy
~/Powerwall-Dashboard/venv/bin/pip install pandas
~/Powerwall-Dashboard/venv/bin/pip install pymysql
~/Powerwall-Dashboard/venv/bin/pip install cryptography
```
13. Re-install docker as per the instructions for docker in the main readme.
14. Bring up the Dashboard `./compose-dash.sh up -d`
15. backfill data using the Tesla History Tool `~/Powerwall-Dashboard/venv/bin/python ~/Powerwall-Dashboard/tools/tesla-history/tesla-history.py --today`

Happy Days.

_Edit: fixed command to backfill history - I'd pasted in the command I use to update ToU data instead from my .bash_history_

**Comment by jasonacox** [comment]
Thanks @BJReplay ! Impressive and helpful detail.
